### PR TITLE
Restore transaction when not saving distance in edit page

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

Restore transaction when not saving distance in edit page
### Fixed Issues

$ https://github.com/Expensify/App/issues/31816
PROPOSAL: https://github.com/Expensify/App/issues/31816#issuecomment-1825535414

### Tests

- New distance request flow
1. Go to FAB > Request money > Distance
2. Enter valid addresses and go to the next step
3. Choose any workspace and go to comfirm page
4. Click on Distance option to edit 
5. Add a stop address that is very far from start and finish address
6. After the error appears, click on back button
7. Verify that the distance is revert to the previous distance
- Edit distance request flow
1. Create a distance 
2. Go to transaction thread report
3. Click on distance to open the edit distance page
4. Edit or add an extra stop address
5. Save the distance
6. Verify that the distance is saved successfully
7. Report steps 3-4 one more time
8. Close the edit distance page without saving
9. Verify that the distance is restored

- [x] Verify that no errors appear in the JS console

### Offline tests

None
### QA Steps

- New distance request flow
1. Go to FAB > Request money > Distance
2. Enter valid addresses and go to the next step
3. Choose any workspace and go to comfirm page
4. Click on Distance option to edit 
5. Add a stop address that is very far from start and finish address
6. After the error appears, click on back button
7. Verify that the distance is revert to the previous distance
- Edit distance request flow
1. Create a distance 
2. Go to transaction thread report
3. Click on distance to open the edit distance page
4. Edit or add an extra stop address
5. Save the distance
6. Verify that the distance is saved successfully
7. Report steps 3-4 one more time
8. Close the edit distance page without saving
9. Verify that the distance is restored
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/a8f625df-238f-4675-a518-11dcb3c2b52e

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/11a51ee4-f859-4981-b6f4-b4fa434f81e1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/958cc116-16e2-4e57-984b-bdfbb1338106

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/3a3154a5-273b-4f69-b59d-459f8fcf1b18

</details>

<details>
<summary>iOS: Native</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/bf2f505f-291f-4390-b9f1-23440ec38e36

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/4f5c9721-c4c4-4d66-9d3e-870f50db2439

</details>

<details>
<summary>iOS: mWeb Safari</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/75b10269-35f9-482d-924a-0d58bd1b8561

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/94e4d2d9-9bcc-4f04-92a1-267291959afb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/05cc7715-fb40-4e69-b9a0-8e0c98159903

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/c0040f59-2053-403f-b9fc-a5271d6f0592

</details>

<details>
<summary>MacOS: Desktop</summary>

- New distance request flow

https://github.com/Expensify/App/assets/129500732/ebfa5abd-fb64-44da-91b3-6b7f2accba66

- Edit distance request flow

https://github.com/Expensify/App/assets/129500732/9f69499e-4904-4a19-8cc8-4a32b68292cb

</details>

### CODE REVIEW

**Review by hoangzinh — APPROVED** [review]
LGTM 👍

File: src/components/DistanceRequest/index.js [hunk 1]
```diff
@@ -238,7 +271,7 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         </>
     );
 
-    if (!isEditing) {
+    if (!isEditingNewRequest) {
```
Comment (neil-marcellini) [inline src/components/DistanceRequest/index.js:274]: I don't understand this change, would you please change it and include a comment in the code?

File: src/components/DistanceRequest/index.js [hunk 2]
```diff
@@ -38,6 +39,9 @@ const propTypes = {
     /** Are we editing an existing distance request, or creating a new one? */
     isEditingRequest: PropTypes.bool,
 
+    /** Are we editing a new distance request */
```
Comment (neil-marcellini) [inline src/components/DistanceRequest/index.js:42]: Can you add more details to this comment to make it clear what this means? I believe it's editing from the confirmation screen when creating a new request. People might think it means editing the waypoints the first time when creating a new distance request.

File: src/components/DistanceRequest/index.js [hunk 3]
```diff
@@ -134,7 +162,7 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
     }, [waypoints, previousWaypoints]);
 
     const navigateBack = () => {
-        Navigation.goBack(isEditing ? ROUTES.MONEY_REQUEST_CONFIRMATION.getRoute(iouType, reportID) : ROUTES.HOME);
+        Navigation.goBack(isEditingNewRequest ? ROUTES.MONEY_REQUEST_CONFIRMATION.getRoute(iouType, reportID) : ROUTES.HOME);
```
Comment (neil-marcellini) [inline src/components/DistanceRequest/index.js:165]: Shouldn't this be `(isEditing || isEditingNewRequest)`?

File: src/pages/iou/NewDistanceRequestPage.js [hunk 4]
```diff
@@ -44,6 +46,7 @@ const defaultProps = {
 // You can't use Onyx props in the withOnyx mapping, so we need to set up and access the transactionID here, and then pass it down so that DistanceRequest can subscribe to the transaction.
 function NewDistanceRequestPage({iou, report, route}) {
     const iouType = lodashGet(route, 'params.iouType', 'request');
+    const isEditing = Navigation.getActiveRoute().includes('address');
```
Comment (neil-marcellini) [inline src/pages/iou/NewDistanceRequestPage.js:49]: NAB: A comment here would be good. Also maybe make the name match the prop name.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looks pretty good, but I have some questions. Could you pleas also test the distance edit flow after creation to make sure there are no regressions?

**Comment by neil-marcellini** [comment]
> @neil-marcellini I tested the edit flow locally before creating the PR. Should we add the test step for edit flow here?

Yes please. We should test flows that might break in relation to our changes, and editing an existing distance request could certainly be effected.

File: src/components/DistanceRequest/index.js (hunk 1 above)
Comment (neil-marcellini) [reply src/components/DistanceRequest/index.js:274]: Ok yes I understand isEditingNewRequest is true when we click the distance field from the confirmation screen while creating a new request.  This condition is deciding whether to return only content, or to wrap it with a screen wrapper and header. We don't need the header for the new distance request page, nor for the editing request page, but we are missing a header when editing a new distance request with the current change 😕 
image

File: src/components/DistanceRequest/index.js (hunk 2 above)
Comment (neil-marcellini) [reply src/components/DistanceRequest/index.js]: Nice thanks

File: src/components/DistanceRequest/index.js (hunk 3 above)
Comment (neil-marcellini) [reply src/components/DistanceRequest/index.js:165]: Oh I see right, when editing a request, and then going back, we just close the modal.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Thanks for the updates. Looks like we need a few more changes. I also found a bug. Please test all related flows with test steps and videos.

File: src/components/DistanceRequest/index.js (hunk 1 above)
Comment (neil-marcellini) [reply src/components/DistanceRequest/index.js:274]: @dukenv0307 please fix this bug

**Review by neil-marcellini — APPROVED** [review]
Ok I tested again and it's working great 🚀 


https://github.com/Expensify/App/assets/26260477/a921bac1-69e7-4465-814f-2264831ad125


https://github.com/Expensify/App/assets/26260477/52d0de0b-d4d3-4a89-9085-fc0ac72a4d8f


https://github.com/Expensify/App/assets/26260477/e087b4e5-a315-455a-a6de-33222cfd084f

**Pull request merged**